MAINT: Throw PdfReadError if Trailer can't be read

ediamondscience opened the pull request:
… Added myself to contributor list.

In reference to #1279 I've modified _reader.py to throw a PdfReadError if the trailer doesn't exist, can't be read, or does not point to the object containing the metadata for the document. 

Three pdf files were uploaded in relation to aforementioned issue, pca_var.pdf contained no trailer, and the other two had corrupted declarations that PyPDF2 could not read. I'll include them here for those interested. 

From Rotary Potentiometer.pdf:
`trailer^M
<</Size 267/Prev 2815238/Root 176 0 R/Info 174 0 R/ID[<8CC42D96833D2A438E62CB60D00F610F><5C480BCE388CB6408C83C951FDD0DA22>]>>^M
startxref^M
0^M
%%EOF^M`

From Status_v1_Reviewers-Guide.pdf:
`trailer`
`/Size`
`447`
`/Root`
`4`
`0`
`R`
`/Info`
`5`
`0`
`R`
`startxref`
`6128518`
(sorry for using multiple code blocks, but I think it's important to show that there are line breaks for each of these lines in my terminal)

We should probably figure out why these trailers are displayed like this. If I had to guess it might be because of a different encoding for line breaks. If we can update the way that we read the trailer to account for this we might be able to read useful information out of more files. 

All three files from the issue should now throw an error when the test code is run. This should close #1279

MartinThoma commented:
Can you please add a test that covers this? You mentioned that https://github.com/p-sampson-II/SEEDLab-TAs-2022/blob/378e6f583ed9de274a8bda9240dfa7aa6872a733/Alex%20Curtis/Datasheets/Rotary%20Potentiometer.pdf shows the issue. You can use the file via:

```python
url = "https://github.com/p-sampson-II/SEEDLab-TAs-2022/blob/378e6f583ed9de274a8bda9240dfa7aa6872a733/Alex%20Curtis/Datasheets/Rotary%20Potentiometer.pdf"
name = "github-Rotary-Potentiometer.pdf"
reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
```

MartinThoma commented on PyPDF2/_reader.py:
```diff
@@ -326,8 +326,11 @@ def metadata(self) -> Optional[DocumentInformation]:
-        return retval
+        if isinstance(obj, type(None)):
+            raise PdfReadError("trailer not found or does not point to document information directory")
+        else:
+            retval.update(obj)  # type: ignore
+            return retval
```
```suggestion
        retval.update(obj)  # type: ignore
        return retval
```

ediamondscience replied:
I'm curious about this one. Why remove the else statement? Is it for performance or something else? Should I have a try statement in place of the if?

MartinThoma commented:
Thank you for your work on this one!

I've just noticed that you've added `resources/unreadablemetadata.pdf` which has 2MB. In order to keep this repository small (and thus easy to clone, also with bad connections), I would like this file to be in https://github.com/py-pdf/sample-files/ .

Could you please add the file there? (I can also do it + use the "Contributed-by" feature of Github to attribute it to you)

MartinThoma commented:
Thank you for your work on this one!

I've just noticed that you've added `resources/unreadablemetadata.pdf` which has 2MB. In order to keep this repository small (and thus easy to clone, also with bad connections), I would like this file to be in https://github.com/py-pdf/sample-files/ .

Could you please add the file there? (I can also do it + use the "Contributed-by" feature of Github to attribute it to you)

ediamondscience commented:
@MartinThoma I've added a pull request to get that file into the correct repository and updated the file location in test_reader.py 

The commit may fail some tests in this repository before the sample files are merged, but I've tested it on my own rig here with pytest and it should work perfectly after the merge.

MartinThoma commented:
I was just blocked at work and added the updated submodule to PyPDF2 :-)

MartinThoma commented:
Could you please do `git rm resources/unreadablemetadata.pdf`?

ediamondscience commented:
Looks like the the reader causes some issues if the metadata is unreadable. Is it the right move to throw a PdfReadError for this situation?

MartinThoma commented:
Sorry, my mistake. I'll fix the test tomorrow 🙈
